Mathlib/Data/Fin/Basic.lean

```diff
@@ -0,0 +1,196 @@
+simp only [Fin.mul_def, Nat.mul_comm]
+
+lemma Fin.add_assoc (a b c : Fin n) : (a + b) + c = a + (b + c) := by
+apply Fin.eq_of_val_eq
+simp only [Fin.mul_def, Fin.add_def, Nat.mod_add_mod, Nat.add_mod_mod, Nat.add_assoc]
+
+instance : Neg (Fin n) where
+  neg := λ a => ⟨(n - a) % n, Nat.mod_lt _ (lt_of_le_of_lt (Nat.zero_le _) a.isLt)⟩
+
+@[simp] lemma Fin.add_left_neg : ∀ (a : Fin n.succ), -a + a = 0
+| ⟨a, isLt⟩ => by
+  apply Fin.eq_of_val_eq
+  simp only [Neg.neg, Fin.add_def, Fin.val_zero]
+  cases a with
+  | zero => rw [Nat.sub_zero, Nat.add_zero, Nat.mod_mod, Nat.mod_self]
+  | succ a =>
+    have h1 : (n.succ - a.succ) % n.succ = (n.succ - a.succ) :=
+      Nat.mod_eq_of_lt (Nat.sub_lt (Nat.succ_pos _) (Nat.succ_pos _))
+    have h_aux : (Nat.succ n - Nat.succ a + Nat.succ a) = Nat.succ n := by
+      rw [Nat.add_comm]
+      exact Nat.add_sub_of_le (Nat.le_of_lt isLt)
+    rw [h1, h_aux, Nat.mod_self]
+
+def Fin.nsmul (x : Nat) (a : Fin n.succ) : Fin n.succ := (Fin.ofNat x) * a
+
+def Fin.gsmul (x : Int) (a : Fin n.succ) : Fin n.succ :=
+  match x with
+  | Int.ofNat x' => Fin.nsmul x' a
+  | Int.negSucc x' => -(Fin.nsmul x'.succ a)
+
+/- Aux lemma that makes nsmul_succ' easier -/
+lemma Fin.nsmuls_eq (x : Nat) : ∀ (a : Fin n.succ), Fin.nsmul x a = Fin.ofNat (x * a.val)
+| ⟨a, isLt⟩ => by
+  apply Fin.eq_of_val_eq
+  simp only [Fin.nsmul, Fin.ofNat, Fin.mul_def]
+  have hh : a % n.succ = a := Nat.mod_eq_of_lt isLt
+  generalize hq : x * a % Nat.succ n = q
+  rw [<- hh, <- Nat.mul_mod, hq]
+
+lemma Fin.nsmul_succ' (x : Nat) (a : Fin n.succ) : Fin.nsmul x.succ a = a + (Fin.nsmul x a) := by
+  simp only [Fin.nsmuls_eq]
+  simp [nsmul, Fin.ofNat, Fin.add_def]
+  exact congrArg (fun x => x % n.succ) (Nat.add_comm (x * a.val) (a.val) ▸ Nat.succ_mul x a.val)
+
+lemma Fin.sub_eq_add_neg (a b : Fin n) : a - b = a + -b := by
+  simp [Fin.add_def, Fin.sub_def, Neg.neg]
+
+@[simp] lemma Fin.gsmul_zero' (a : Fin n.succ) : Fin.gsmul 0 a = (0 : Fin n.succ) := by
+  simp only [Fin.gsmul, Fin.nsmul, Fin.of_nat_zero, Fin.zero_mul]
+
+lemma Fin.gsmul_succ' (m : ℕ) (a : Fin n.succ) : Fin.gsmul (Int.ofNat m.succ) a = a + Fin.gsmul (Int.ofNat m) a := by
+  simp only [Fin.gsmul, Fin.nsmul_succ']
+
+lemma Fin.gsmul_neg' (m : ℕ) (a : Fin n.succ) : gsmul (Int.negSucc m) a = -(gsmul (m.succ) a) := by
+  simp only [Fin.gsmul, Fin.nsmul]
+
+instance : AddSemigroup (Fin n.succ) where
+  add_assoc := Fin.add_assoc
+
+instance : AddMonoid (Fin n.succ) where
+  add_zero := Fin.add_zero
+  zero_add := Fin.zero_add
+  nsmul := Fin.nsmul
+  nsmul_zero' := Fin.zero_mul
+  nsmul_succ' := Fin.nsmul_succ'
+
+instance : SubNegMonoid (Fin n.succ) where
+  sub_eq_add_neg := Fin.sub_eq_add_neg
+  gsmul := Fin.gsmul
+  gsmul_zero' := Fin.gsmul_zero'
+  gsmul_succ' := Fin.gsmul_succ'
+  gsmul_neg' := Fin.gsmul_neg'
+
+instance : AddGroup (Fin n.succ) where
+  add_left_neg := Fin.add_left_neg
+
+def Fin.overflowingAdd (a b : Fin n) : (Bool × Fin n) := (n <= a.val + b.val, a + b)
```

Comment:
We don't have these in mathlib, are these useful?  (I can somewhat see the point in checkedAdd, but this definition seems strictly less useful than just writing `n <= a.val + b.val`.)

Reply:
I think so; I'm not super attached to this particular interface, but I think putting overflow/underflow checks behind a definition is preferable to having users do the checks themselves. It's easy to make stupid mistakes with inequalities, and Fin has a second sharp edge in that `n <= a + b` is wrong, but tempting if you're not thinking very hard.

Reply:
My point was rather that a definition like `Fin.isNonOverflowingAdd (a b : Fin n) : Bool` might be more ergonomic because pattern matches are always awkward in proofs, and writing `(overflowingAdd a b).1` gets old.

But I won't insist on anything here.  The standard disclaimer in the readme applies.

Reply:
That's a good point. I'll add a boolean check like you described and people can feel out what idioms end up being more comfortable. I suspect you're right that when proofs are in the mix, the use case of "I just want to know whether these overflow" is more common than "I want the result, and I want to know whether they overflow".